Source/Core/VideoBackends/D3D12/D3DBase.cpp

```diff
@@ -0,0 +1,924 @@
+	current_command_list = nullptr;
+
+	// unload DLLs
+	UnloadDXGI();
+	UnloadD3DCompiler();
+	UnloadD3D();
+}
+
+const std::string VertexShaderVersionString()
+{
+	return "vs_5_0";
+}
+
+const std::string GeometryShaderVersionString()
+{
+	return "gs_5_0";
+}
+
+const std::string PixelShaderVersionString()
+{
+	return "ps_5_0";
+}
+
+D3DTexture2D* &GetBackBuffer()
+{
+	return s_backbuf[s_current_back_buf];
+}
+
+unsigned int GetBackBufferWidth()
+{
+	return s_xres;
+}
+
+unsigned int GetBackBufferHeight()
+{
+	return s_yres;
+}
+
+// Returns the maximum width/height of a texture.
+unsigned int GetMaxTextureSize()
+{
+	return D3D11_REQ_TEXTURE2D_U_OR_V_DIMENSION;
+}
+
+void Reset()
+{
+	command_list_mgr->ExecuteQueuedWork(true);
+
+	// release all back buffer references
+	for (UINT i = 0; i < ARRAYSIZE(s_backbuf); i++)
+	{
+		SAFE_RELEASE(s_backbuf[i]);
+	}
+
+	D3D::command_list_mgr->ImmediatelyDestroyAllResourcesScheduledForDestruction();
+
+	// resize swapchain buffers
+	RECT client;
+	GetClientRect(hWnd, &client);
+	s_xres = client.right - client.left;
+	s_yres = client.bottom - client.top;
+
+	CheckHR(s_swapchain->ResizeBuffers(SWAP_CHAIN_BUFFER_COUNT, s_xres, s_yres, DXGI_FORMAT_R8G8B8A8_UNORM, 0));
+
+	// recreate back buffer textures
+
+	HRESULT hr = S_OK;
+	ID3D12Resource* buf12 = nullptr;
+
+	for (UINT i = 0; i < SWAP_CHAIN_BUFFER_COUNT; i++)
+	{
+		ID3D12Resource* buf12;
+		hr = s_swapchain->GetBuffer(i, IID_PPV_ARGS(&buf12));
+
+		CHECK(SUCCEEDED(hr), "Create back buffer texture");
+
+		s_backbuf[i] = new D3DTexture2D(buf12,
+			D3D11_BIND_RENDER_TARGET,
+			DXGI_FORMAT_UNKNOWN,
+			DXGI_FORMAT_UNKNOWN,
+			DXGI_FORMAT_UNKNOWN,
+			false,
+			D3D12_RESOURCE_STATE_PRESENT
+			);
+
+		CHECK(s_backbuf != nullptr, "Create back buffer texture");
+
+		SAFE_RELEASE(buf12);
+		SetDebugObjectName12(s_backbuf[i]->GetTex12(), "backbuffer texture");
+	}
+
+	s_current_back_buf = 0;
+
+	s_backbuf[s_current_back_buf]->TransitionToResourceState(current_command_list, D3D12_RESOURCE_STATE_RENDER_TARGET);
+}
+
+bool BeginFrame()
+{
+	if (s_frame_in_progress)
+	{
+		PanicAlert("BeginFrame called although a frame is already in progress");
+		return false;
+	}
+	s_frame_in_progress = true;
+	return (device12 != nullptr);
+}
+
+void EndFrame()
+{
+	if (!s_frame_in_progress)
+	{
+		PanicAlert("EndFrame called although no frame is in progress");
+		return;
+	}
+	s_frame_in_progress = false;
+}
+
+void Present()
+{
+	// Persist last present QPC across function calls, used to make sure we don't get throttled by Presenting too much.
+	static LARGE_INTEGER s_last_present;
+
+	LARGE_INTEGER current_timestamp;
+	LARGE_INTEGER frequency;
+	QueryPerformanceCounter(&current_timestamp);
+	QueryPerformanceFrequency(&frequency);
+
+	UINT present_flags = DXGI_PRESENT_TEST; // DXGI_PRESENT_TEST means do not actually present frame.
+
+	if ((UINT)g_ActiveConfig.IsVSync() || (((current_timestamp.QuadPart - s_last_present.QuadPart) * 1000) / frequency.QuadPart >= (16.667 / 2)))
```

Comment:
I'll try the GetFrameLatencyWaitableObject, it should be cleaner than this approach (a low bar). I'll have to perform the check on the background thread unfortunately, since we're not actually Presenting here..